Source/WarpX.H

```diff
@@ -946,6 +996,9 @@ private:
 #   endif
 #endif
 
+     // Warning manager: it allows recording and printing error messages
+    std::unique_ptr<Utils::WarnManager> m_p_warn_manager;
```

Comment:
Does it make sense to move this out of the WarpX class and into its own namespace, like `Warn` or something? Then could do `Warn::RecordWarning` instead of having to do `WarpX::GetInstance()::RecordWarning`. You could set it up / tear it down with `Warn::Initialize()` and `Warn::Finalize()` - this is the pattern used, for example, by AMReX [Arena](https://github.com/AMReX-Codes/amrex/blob/development/Src/Base/AMReX_Arena.cpp#L32)

Reply:
I think that it makes sense. WarpX class is already very complex. Moreover, having `RecordWarning` as a member function of WarpX forces to include `WarpX.H` in some files,  and `WarpX.H` is not exactly a lightweight header file.

Reply:
As I mentioned few days ago, I like the idea of placing `RecordWarning` out of `WarpX.H`, as suggested  by @atmyers .
Concerning the implementation details, do you have preferences between a global object and a singleton? I think that C++ guidelines discourage both, but they also state that a global object is often better than a singleton.